model/certificate_test.go

```diff
@@ -0,0 +1,154 @@
+package model
+
+import (
+	"testing"
+	"time"
+)
+
+type certTest struct {
+	testName      string
+	domains       []string
+	email         string
+	expectedError string
+}
+
+func TestNewCertificate(t *testing.T) {
+	certTests := []certTest{
+		{
+			"happy path",
+			[]string{"example.com", "example2.com"},
+			"test@notexample.com",
+			"",
+		},
+		{
+			"no domains",
+			[]string{},
+			"test@notexample.com",
+			ErrInvalidDomains.Error(),
+		},
+		{
+			"email at example.com",
+			[]string{"example.com"},
+			"test@example.com",
+			"acme: error: 400 :: POST :: https://acme-v02.api.letsencrypt.org/acme/new-acct :: urn:ietf:params:acme:error:invalidEmail :: Error creating new account :: invalid contact domain. Contact emails @example.com are forbidden, url: ",
+		},
+	}
+
+	for _, ct := range certTests {
+		t.Run(ct.testName, func(t *testing.T) {
+			c, err := NewCertificate(ct.domains, ct.email)
+
+			if err == nil {
+				if ct.expectedError != "" {
+					t.Error("no error returned when expected")
+				}
+			}
+
+			if err != nil {
+				if err.Error() != ct.expectedError {
+					t.Errorf("error mismatch: got %s, expected %s\n", err.Error(), ct.expectedError)
+				}
+				if c != nil {
+					t.Error("certificate should be nil on error")
+				}
+				return
+			}
+
+			if c.ID == "" {
+				t.Error("certificate ID blank")
+			}
+
+			if c.Secret == "" {
+				t.Error("certificate Secret blank")
+			}
+
+			match := testEq(ct.domains, c.Domains)
+			if !match {
+				t.Error("given domains and certificate domains mismatch")
+			}
+
+			if c.CommonName != ct.domains[0] {
+				t.Error("common name is not correct domain")
+			}
+
+			if c.CertURL != "" {
+				t.Error("cert url should be blank")
+			}
+			if c.CertStableURL != "" {
+				t.Error("cert stable url should be blank")
+			}
+
+			if len(c.PrivateKey) != 0 {
```

Comment:
I assume this passes and I'm missing something, but it looks like `PrivateKey` is filled out by the `NewCertificate` function. So shouldn't this always fail?

Reply:
There's a small difference between the ACMEKey (private key) and the actual Certificate's Private Key. This isn't super clear in the actual definition of NewCertificate.

Reply:
Oh whoops I see it now. I must have switched the fieldname fieldvalue order around in my head lol